### PR TITLE
Add sidewalk method in alignment

### PR DESCRIPTION
**Pull request by benjello**
I try to add the method implemented by @AlexisEidelman.
To test it the method, i duplicated the test_align_abs. But there are errors. Insights welcome.
@gdementen: can't we have a the error message to show the concerned expression ?

### CODE REVIEW

File: liam2/alignment.py [hunk 1]
```diff
@@ -115,11 +115,21 @@ def align_get_indices_nd(ctx_length, groups, need, filter_value, score,
...
+                        # TODO: use np.partition (np1.8+)
+                        sorted_local_indices = np.argsort(maybe_members_rank_value)
+                        sorted_global_indices = \
+                            group_maybe_indices[sorted_local_indices]
+                    if method == 'sidewalk':
+                        score_max = max(score[group_maybe_indices])
+                        score_min = min(score[group_maybe_indices])
+                        if  score_max > 1 or score_min < 0:
+                            raise Exception("""
+Score values are in the interva {} - {}.
```
Comment (gdementen) [inline liam2/alignment.py]: interval typo

File: liam2/alignment.py [hunk 2]
```diff
@@ -115,11 +115,21 @@ def align_get_indices_nd(ctx_length, groups, need, filter_value, score,
...
-                        group_maybe_indices[sorted_local_indices]
+                    if method == 'default':
+                        maybe_members_rank_value = score[group_maybe_indices]
+                        # TODO: use np.partition (np1.8+)
+                        sorted_local_indices = np.argsort(maybe_members_rank_value)
+                        sorted_global_indices = \
+                            group_maybe_indices[sorted_local_indices]
+                    if method == 'sidewalk':
+                        score_max = max(score[group_maybe_indices])
+                        score_min = min(score[group_maybe_indices])
```
Comment (gdementen) [inline liam2/alignment.py]: It would probably be better to do the check on the whole score vector, not on each group.

File: liam2/alignment.py [hunk 3]
```diff
@@ -128,8 +138,18 @@ def align_get_indices_nd(ctx_length, groups, need, filter_value, score,
...
                 maybe_to_take = affected - num_always
-                # take the last X individuals (ie those with the highest score)
-                indices_to_take = sorted_global_indices[-maybe_to_take:]
+                if method == 'default':
+                    # take the last X individuals (ie those with the highest score)
+                    indices_to_take = sorted_global_indices[-maybe_to_take:]
+                elif method == 'sidewalk':
+                    assert maybe_to_take <= sum(score[sorted_global_indices]), \
+                        "Can't use Sidewalk with need > sum of probabilities"
+                    U = np.random.uniform() + np.arange(maybe_to_take)
```
Comment (gdementen) [inline liam2/alignment.py]: lowercase u, please

File: liam2/alignment.py [hunk 4]
```diff
@@ -128,8 +138,18 @@ def align_get_indices_nd(ctx_length, groups, need, filter_value, score,
...
+                    # on the random sample, score are cumulated and then, we extract indices
+                    # of each value before each value of U
+                    indices_to_take = np.searchsorted(np.cumsum(score[sorted_global_indices]), U)
+                    indices_to_take = sorted_global_indices[indices_to_take]
+                    boum
```
Comment (gdementen) [inline liam2/alignment.py]: boum?

File: liam2/alignment.py [hunk 5]
```diff
@@ -295,11 +317,26 @@ def _display_unaligned(self, expressions, ids, columns, unaligned):
 
     def compute(self, context, score, need, filter=None, take=None, leave=None,
                 expressions=None, possible_values=None, errors='default',
-                frac_need='uniform', link=None, secondary_axis=None):
+                frac_need='uniform', link=None, secondary_axis=None,
+                method='default'):
```
Comment (gdementen) [inline liam2/alignment.py]: the default method should not be called "default", but "bysorting" or "sort" or something more descriptive like that

File: liam2/alignment.py [hunk 6]
```diff
@@ -295,11 +317,26 @@ def _display_unaligned(self, expressions, ids, columns, unaligned):
...
                 expressions=None, possible_values=None, errors='default',
-                frac_need='uniform', link=None, secondary_axis=None):
+                frac_need='uniform', link=None, secondary_axis=None,
+                method='default'):
+
+        if method not in ("default", "sidewalk"):
+            raise Exception("Method for alignment should be either 'default' "
+                            "either 'sidewalk'")
+        if need is None and method != 'sidewalk':
+            raise Exception("No default value for need if method is not sidewalk")
```
Comment (gdementen) [inline liam2/alignment.py]: if need can be None for sidewalk, maybe need should be made optional (None by default)

File: liam2/alignment.py [hunk 7]
```diff
@@ -295,11 +317,26 @@ def _display_unaligned(self, expressions, ids, columns, unaligned):
...
+            # Note: need is calculated over score and we could think of
+            # calculate without leave_filter and without take_filter
+            if need is None:
+                need = sum(score)
+
         # need is a single scalar
         # if not isinstance(need, (tuple, list, np.ndarray)):
         if np.isscalar(need):
             need = [need]
+             # need = [np.floor(need)]
```
Comment (gdementen) [inline liam2/alignment.py]: scrap this

File: liam2/alignment.py [hunk 8]
```diff
@@ -329,15 +366,17 @@ def compute(self, context, score, need, filter=None, take=None, leave=None,
 
         func = self.align_no_link if link is None else self.align_link
         return func(context, score, need, filter, take, leave, expressions,
-                    possible_values, errors, frac_need, link, secondary_axis)
+                    possible_values, errors, frac_need, link, secondary_axis, method = method)
```
Comment (gdementen) [inline liam2/alignment.py]: in this case, I'd prefer, "secondardy_axis, method)" (ie no method=)

File: liam2/alignment.py [hunk 9]
```diff
@@ -536,8 +576,12 @@ def __init__(self, score=None, proportions=None,
                                 "arguments should have the same length")
 
         if proportions is None and fname is None:
-            raise Exception("align() needs either an fname or proportions "
-                            "arguments")
+            if method == 'default':
+                raise Exception("align() needs either an fname or proportions "
+                                "arguments")
+            if method == 'sidewalk':
```
Comment (gdementen) [inline liam2/alignment.py]: elif?

File: liam2/alignment.py [hunk 10]
```diff
@@ -547,7 +591,8 @@ def __init__(self, score=None, proportions=None,
         super(Alignment, self).__init__(score, proportions,
                                         filter, take, leave,
                                         expressions, possible_values,
-                                        errors, frac_need)
+                                        errors, frac_need,
+                                        method = method)
```
Comment (gdementen) [inline liam2/alignment.py]: no "method="

File: liam2/alignment.py [hunk 11]
```diff
@@ -558,4 +603,4 @@ def _get_need_correction(self, groups, possible_values):
 functions = {
     'align_abs': AlignmentAbsoluteValues,
     'align': Alignment
-}
+    }
```
Comment (gdementen) [inline liam2/alignment.py]: revert this

**Comment by gdementen** [comment]
please also do not commit the changes to the .h5 files!

**Comment by gdementen** [comment]
@benjello about the exception, it seems clear to me (the test is bogus: you give age as score, which is obviously not between 0 and 1).

- int_aligned: align_abs(age, ...

exception:
Score values are in the interva 0 - 95.

Sidewalk method can be used only with a score between 0 and 1. You may want to use a logistic function

File: liam2/alignment.py (hunk 8 above)
Comment (benjello) [inline liam2/alignment.py]: What is the policy ? The function has a lot of arguments and it took me some time to trace the error. Just asking though, you are the boss ;-)

**Comment by benjello** [comment]
@gdementen : thanks for you quick reply for this messy PR.
For the exception, the message is shorter on my side (and on travis) and it is not very clear which expression is to blame.

**Comment by benjello** [comment]
Closing opening a new cleaner one with some problems unsolved.

**Comment by gdementen** [comment]
@benjello for the exception message, I got it from travis!

**Pull request closed without merging**